File: app/src/main/java/thevoiceless/setproxy/adapters/ProxyConfigurationAdapter.java
package thevoiceless.setproxy.adapters;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import thevoiceless.setproxy.R;
import thevoiceless.setproxy.data.ProxyConfiguration;

import static butterknife.ButterKnife.findById;

/**
 * Created by riley on 11/27/15.
 */
public class ProxyConfigurationAdapter extends RecyclerView.Adapter<ProxyConfigurationAdapter.ProxyViewHolder> {

    public interface OnItemClickListener {
        void onItemClick(@NonNull final View v, final int position);
    }

    public static class ProxyViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        public int position;
        public TextView host;
        public TextView port;
        public OnItemClickListener listener;

        public ProxyViewHolder(View itemView) {
            super(itemView);

            host = findById(itemView, R.id.proxy_host);
            port = findById(itemView, R.id.proxy_port);

            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            if (listener != null) listener.onItemClick(v, position);
        }
    }

    private ArrayList<ProxyConfiguration> mProxies;
    private OnItemClickListener mListener;

    public ProxyConfigurationAdapter() {
        mProxies = new ArrayList<>();
    }

    public ProxyConfigurationAdapter(@NonNull final List<ProxyConfiguration> proxies) {
        mProxies = new ArrayList<>(proxies);
    }

    @Override
    public ProxyViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_proxy, parent, false);
        return new ProxyViewHolder(v);
    }

    @Override
    public void onBindViewHolder(ProxyViewHolder holder, int position) {
        ProxyConfiguration proxy = mProxies.get(position);
        holder.position = position;
        holder.host.setText(proxy.getHost());
        holder.port.setText(proxy.getPort());
        holder.listener = mListener;
    }

    public ProxyConfiguration getItem(final int position) {
        return mProxies.get(position);
    }

    public void setOnItemClickListener(@Nullable final OnItemClickListener listener) {
        mListener = listener;
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return mProxies.size();
    }

    public void addProxy(@NonNull final ProxyConfiguration proxy) {
        mProxies.add(0, proxy);
        notifyItemInserted(0);
    }

    public void setData(@NonNull final List<ProxyConfiguration> proxies) {
        mProxies = new ArrayList<>(proxies);
        notifyDataSetChanged();
    }
}

File: app/src/main/java/thevoiceless/setproxy/adapters/ProxyConfigurationAdapter.kt
package thevoiceless.setproxy.adapters

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import java.util.ArrayList

import thevoiceless.setproxy.R
import thevoiceless.setproxy.data.ProxyConfiguration

/**
 * Created by riley on 11/27/15.
 */
class ProxyConfigurationAdapter : RecyclerView.Adapter<ProxyConfigurationAdapter.ProxyViewHolder>() {

    private val proxies = mutableListOf<ProxyConfiguration>()
    private var clickListener: OnItemClickListener? = null

    interface OnItemClickListener {
        fun onItemClick(v: View, proxy: ProxyConfiguration)
    }

    class ProxyViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        var host: TextView = itemView.findViewById(R.id.proxy_host)
        var port: TextView = itemView.findViewById(R.id.proxy_port)

        var listener: OnItemClickListener? = null
        lateinit var proxy: ProxyConfiguration

        init {
            itemView.setOnClickListener(this)
        }

        override fun onClick(v: View) {
            listener?.onItemClick(v, proxy)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProxyViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.list_item_proxy, parent, false)
        return ProxyViewHolder(v)
    }

    override fun onBindViewHolder(holder: ProxyViewHolder, position: Int) {
        val proxy = proxies[position]
        holder.apply {
            this.proxy = proxy
            host.text = proxy.host
            port.text = proxy.port
            listener = clickListener
        }
    }

    fun setOnItemClickListener(listener: OnItemClickListener?) {
        clickListener = listener
        notifyDataSetChanged()
    }

    override fun getItemCount() = proxies.size

    fun addProxy(proxy: ProxyConfiguration) {
        proxies.add(0, proxy)
        notifyItemInserted(0)
    }

    fun setData(proxies: List<ProxyConfiguration>) {
        this.proxies.apply {
            clear()
            addAll(proxies)
        }
        notifyDataSetChanged()
    }
}
